File: custom_dir/custom_actions/count_action.py
import json
import time

# -*- coding: UTF-8 -*-
from maa.context import Context
from maa.custom_action import CustomAction

from maa.agent.agent_server import AgentServer

@AgentServer.custom_action("CountAction")
class CountAction(CustomAction):
    def __init__(self):
        super().__init__()
        self._should_stop = False

    def run(self,
            context: Context,
            argv: CustomAction.RunArg) -> bool:
        """
        执行计数动作，当达到目标计数时执行next任务

        :param context: 运行上下文
        :param argv: 运行参数，包括next_task、count_target和interval
        :return: 是否执行成功
        """
        # 重置停止标志
        self._should_stop = False

        # 读取 custom_param 的参数：{"next_task": "TaskName", "count_target": x, "interval": y, "task_to_count": "TaskToCount"}
        json_data = json.loads(argv.custom_action_param)

        # 获取参数
        next_task = json_data.get("next_task", "")
        count_target = json_data.get("count_target", 1)
        interval = json_data.get("interval", 1.0)  # 默认间隔1秒
        task_to_count = json_data.get("task_to_count", "")

        # 验证参数有效性
        if not next_task or count_target < 1 or not task_to_count:
            print(f"无效的参数: next_task={next_task}, count_target={count_target}, task_to_count={task_to_count}")
            return False

        print(f"开始计数动作: 目标={count_target}, 间隔={interval}秒, 计数任务={task_to_count}, 下一任务={next_task}")

        # 当前计数
        current_count = 0

        # 开始计数循环
        while current_count < count_target and not self._should_stop:
            print(f"当前计数: {current_count}/{count_target}")
            image = context.tasker.controller.post_screencap().wait().get()
            # 执行需要计数的任务
            result = context.run_recognition(task_to_count,image)
            # 检查任务执行结果
            if result and result.best_result:
                current_count += 1
                print(f"任务 {task_to_count} 执行成功，计数增加到 {current_count}")
            else:
                print(f"任务 {task_to_count} 执行失败，计数保持 {current_count}")

            # 如果达到目标计数，退出循环
            if current_count >= count_target:
                break

            # 等待指定的间隔时间
            if not self._should_stop and interval > 0:
                print(f"等待 {interval} 秒...")
                time.sleep(interval)

        # 如果因为达到目标退出循环，则执行next任务
        if current_count >= count_target and not self._should_stop:
            print(f"达到目标计数 {count_target}，执行下一任务: {next_task}")
            next_result = context.run_task(next_task)
            print(f"下一任务 {next_task} 执行结果: {'成功' if next_result else '失败'}")
            return True
        elif self._should_stop:
            print("计数动作被手动停止")
            return False
        else:
            print("计数未达到目标但循环已结束")
            return False

    def stop(self) -> None:
        """停止执行的逻辑"""
        print("收到停止计数的请求")
        self._should_stop = True

File: agent/actions/count_action/count_action.go
package count_action

import (
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/MaaXYZ/maa-framework-go/v4"
)

type CountAction struct {
	shouldStop bool
	mu         sync.Mutex
}

type CountParams struct {
	NextTask    string  `json:"next_task"`
	CountTarget int     `json:"count_target"`
	Interval    float64 `json:"interval"`
	TaskToCount string  `json:"task_to_count"`
}

// Run 执行计数动作，当达到目标计数时执行next任务
func (a *CountAction) Run(ctx *maa.Context, arg *maa.CustomActionArg) bool {
	// 重置停止标志
	a.mu.Lock()
	a.shouldStop = false
	a.mu.Unlock()

	var params CountParams
	if err := json.Unmarshal([]byte(arg.CustomActionParam), &params); err != nil {
		fmt.Printf("解析参数失败: %v\n", err)
		return false
	}

	// 设置默认值
	if params.CountTarget < 1 {
		params.CountTarget = 1
	}
	if params.Interval == 0 {
		params.Interval = 1.0
	}

	// 验证参数
	if params.NextTask == "" || params.TaskToCount == "" {
		fmt.Printf("无效的参数: next_task=%s, count_target=%d, task_to_count=%s\n",
			params.NextTask, params.CountTarget, params.TaskToCount)
		return false
	}

	fmt.Printf("开始计数动作: 目标=%d, 间隔=%.1f秒, 计数任务=%s, 下一任务=%s\n",
		params.CountTarget, params.Interval, params.TaskToCount, params.NextTask)

	currentCount := 0

	// 开始计数循环
	for currentCount < params.CountTarget {
		a.mu.Lock()
		stop := a.shouldStop
		a.mu.Unlock()
		if stop {
			break
		}

		fmt.Printf("当前计数: %d/%d\n", currentCount, params.CountTarget)

		// 截图并执行识别
		controller := ctx.GetTasker().GetController()
		if controller == nil {
			fmt.Println("获取控制器失败")
			return false
		}
		controller.PostScreencap().Wait()
		img, err := controller.CacheImage()
		if err != nil {
			fmt.Printf("获取截图失败: %v\n", err)
			return false
		}
		result, err := ctx.RunRecognition(params.TaskToCount, img, nil)
		if err != nil {
			fmt.Printf("识别任务 %s 执行失败: %v\n", params.TaskToCount, err)
		}

		if result != nil && result.Hit {
			currentCount++
			fmt.Printf("任务 %s 执行成功，计数增加到 %d\n", params.TaskToCount, currentCount)
		} else {
			fmt.Printf("任务 %s 执行失败，计数保持 %d\n", params.TaskToCount, currentCount)
		}

		// 如果达到目标计数，退出循环
		if currentCount >= params.CountTarget {
			break
		}

		// 等待指定的间隔时间
		a.mu.Lock()
		stop = a.shouldStop
		a.mu.Unlock()
		if !stop && params.Interval > 0 {
			fmt.Printf("等待 %.1f 秒...\n", params.Interval)
			time.Sleep(time.Duration(params.Interval * float64(time.Second)))
		}
	}

	// 检查是否因为达到目标退出循环
	a.mu.Lock()
	stop := a.shouldStop
	a.mu.Unlock()

	if currentCount >= params.CountTarget && !stop {
		fmt.Printf("达到目标计数 %d，执行下一任务: %s\n", params.CountTarget, params.NextTask)
		_, err := ctx.RunTask(params.NextTask)
		if err != nil {
			fmt.Printf("执行下一任务 %s 失败: %v\n", params.NextTask, err)
			return false
		}
		return true
	} else if stop {
		fmt.Println("计数动作被手动停止")
		return false
	}

	fmt.Println("计数未达到目标但循环已结束")
	return false
}